packages/itwinui-react/src/core/Menu/MenuItem.tsx

```diff
@@ -139,112 +201,139 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
       case 'Enter':
       case ' ':
       case 'Spacebar': {
-        !disabled && onClick?.(value);
+        onClick();
         event.preventDefault();
         break;
       }
-      case 'ArrowRight': {
-        if (subMenuItems.length > 0) {
-          setIsSubmenuVisible(true);
+      default:
+        break;
+    }
+  };
 
-          // flush and reset state so we are ready to focus again next time
-          flushSync(() => setFocusOnSubmenu(true));
-          setFocusOnSubmenu(false);
+  const onMouseEnter = (e: React.MouseEvent<HTMLElement>) => {
+    // Focus the item when hovered.
+    if (e.target === e.currentTarget) {
+      menuItemRef.current?.focus();
 
-          event.preventDefault();
-          event.stopPropagation();
-        }
-        break;
+      // Since we manually focus the MenuItem on hover, we need to manually update the active index for
+      // Floating UI's keyboard navigation to work correctly.
+      if (parent != null && indexInParentTree != null) {
+        parent.setActiveIndex(indexInParentTree);
       }
```

Comment:
@mayank99 and I had removed this in a previous commit (https://github.com/iTwin/iTwinUI/pull/1942/commits/3861266a09f6d06d6a5f79e6e77261b56b12f309). But I brought it back since I now remember why I had it in the first place. It was because if we don't update the active index upon hover/focus, then Floating UI's `useListNavigation` keeps the stale `activeIndex`.

<details>
<summary>Screen recording</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/29fb5fc2-3ce0-4761-991f-6800f6edca9b

</details>

One small difference is that instead of updating the parent's activeIndex in `onFocus`, I am now doing it in `onMouseEnter`, where it is really needed because that is where we're manually changing the focus instead of letting `useListNavigation` handle it.

Reply:
I realized the I was not using the `listNavigation` interaction in the `Menu` in `DropdownMenu`. Thus, the previously mentioned issue existed in the first-level `Menu` in the `DropdownMenu`.

<details>
<summary>Screen recording</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/bba7f760-a8ab-410e-b1e4-4589b1213fc1

</details>

This is now fixed in [`8a1bc78`](https://github.com/iTwin/iTwinUI/pull/1942/commits/8a1bc78647fe9c7d49a26f54989084279c858360). Since there are many contexts, open to suggestions (if any) for context names. In a follow up PR, I can try using [`jotai`](https://jotai.org/) to reduce the number of contexts, similar to https://github.com/iTwin/iTwinUI/pull/2011.

Reply:
Does this change mean you can remove the keyboard navigation logic from `Menu`?

Reply:
>  Since there now are many contexts

I'm not sure I understand why the contexts needed to be split? `setHasFocusedNodeInSubmenu` could be made optional instead?

Reply:
> Does this change mean you can remove the keyboard navigation logic from `Menu`?

Yes, it can be removed if the children of `Menu` is always going to be a `MenuItem`. This is because `MenuItem` handles setting the parent's `listItemsRef` ([code](https://github.com/iTwin/iTwinUI/pull/1942/files#diff-4410321a0ef9f99d8d2c38b7907baf2e10f7fdf8efcff3a9ca5245db96d35db9R270-R275)) which is needed for Floating UI's`useListNavigation` to work.

But since `Menu` supports non-`MenuItem` children too, I guess some cloning will be required to add a `ref` function that updates `listItemsRef`. So, we would need to add this to all internal uses of the `Menu` component and add this cloning logic to the children inside the `Menu`.

Since some more investigation might be required, I think this can be done after this PR. If so, I can add this as an after PR TODO.

Reply:
> But since Menu supports non-MenuItem children too

This is a contrived case that will never happen.

> I think this can be done after this PR

I'm ok with this, but it should be done sooner rather than later, since we now have duplicated logic for keyboard navigation.

Reply:
> This is a contrived case that will never happen.

I see, then should we update the JSDocs in a few places? Because it gives the notion that non-`MenuItem` are fully supported/allowed. Examples:

https://github.com/iTwin/iTwinUI/blob/d1af8d045f7117476abc06467d98b314b92a88c5/packages/itwinui-react/src/core/Menu/Menu.tsx#L17

https://github.com/iTwin/iTwinUI/blob/d1af8d045f7117476abc06467d98b314b92a88c5/packages/itwinui-react/src/core/ComboBox/ComboBox.tsx#L120

https://github.com/iTwin/iTwinUI/blob/d1af8d045f7117476abc06467d98b314b92a88c5/packages/itwinui-react/src/core/Select/Select.tsx#L514

> I'm ok with this, but it should be done sooner rather than later

Sounds good, added an after PR TODO with the prefix "**(Important)**" to prioritize it.

Reply:
`ComboBox` and `Select` (public) have completely different semantics from `Menu` (private), so they should not even be using `MenuItem` in the first place. It makes more sense to use [`ListItem`](https://itwinui.bentley.com/docs/list#interactive-states).

No jsdoc changes necessary in this PR, but here's what I think makes the most sense as a plan for near-future:
1. Remove redundant logic from `Menu` and abstract some parts of `DropdownMenu`/`MenuItem` into `Menu`.
2. Replace incorrect usages of `Menu` (mainly in `ComboBox` and `Select`) with a new private component (could be called `Listbox`).
3. Deprecate `MenuItem` and replace with something like `DropdownMenu.Item` (and maybe `ComboBox.Item`/`Select.Item`).
4. Update all documentation to show the correct usage of all components.